sdk/dotnet/Pulumi/Serialization/Serializer.cs

```diff
@@ -137,17 +140,49 @@ public Serializer(bool excessiveDebugOutput)
 
                 var data = await output.GetDataAsync().ConfigureAwait(false);
                 this.DependentResources.AddRange(data.Resources);
+                var propResources = new HashSet<Resource>();
+                propResources.UnionWith(data.Resources);
+
 
                 // When serializing an Output, we will either serialize it as its resolved value or the "unknown value"
                 // sentinel. We will do the former for all outputs created directly by user code (such outputs always
                 // resolve isKnown to true) and for any resource outputs that were resolved with known values.
                 var isKnown = data.IsKnown;
                 var isSecret = data.IsSecret;
 
+                var child = new Serializer(_excessiveDebugOutput, keepOutputValues: false);
+                var value = await child.SerializeAsync($"{ctx}.id", data.Value, keepResources).ConfigureAwait(false);
+                var promiseDeps = child.DependentResources;
+                this.DependentResources.UnionWith(promiseDeps);
+
+                if (_keepOutputValues /* && monitor supports output values */)
+                {
+                    // do thing
+                    var urnDeps = new HashSet<Resource>();
+                    foreach (var resource in propResources)
+                    {
+                        var childURNResolver = new Serializer(_excessiveDebugOutput, keepOutputValues: false);
+                        await childURNResolver.SerializeAsync($"{ctx} dependency", resource.Urn, keepResources);
+                        urnDeps.UnionWith(childURNResolver.DependentResources);
+                    }
+                    propResources.UnionWith(urnDeps);
+                    this.DependentResources.UnionWith(urnDeps);
+
+                    var depencdencies = await Deployment.GetAllTransitivelyReferencedResourceUrnsAsync(propResources);
```

Comment:
Do we need the transitive closure here?  Is that how the other SDKs do it?

I also wonder if this leads to hangs with cycles, exacerbating https://github.com/pulumi/pulumi/issues/7900

Reply:
This is how the other SDKs do it: https://github.com/pulumi/pulumi/blob/e629bc32d5afe9615869060235b907fe6e92f0b2/sdk/nodejs/runtime/rpc.ts#L386. `GetAllTransitivelyReferencedResourceUrnsAsync` should handle cycles: https://github.com/pulumi/pulumi/blob/e629bc32d5afe9615869060235b907fe6e92f0b2/sdk/dotnet/Pulumi/Deployment/Deployment_Prepare.cs#L183